--- bits/alltypes.h
+++ bits/alltypes.h
@@ -299,17 +299,32 @@
 #endif

 #if defined(__NEED_pthread_mutexattr_t) && !defined(__DEFINED_pthread_mutexattr_t)
-typedef struct { unsigned __attr; } pthread_mutexattr_t;
+typedef struct { union { unsigned __attr;
+#ifdef __aarch64__
+	long __glibc_compat;
+#endif
+};
+} pthread_mutexattr_t;
 #define __DEFINED_pthread_mutexattr_t
 #endif

 #if defined(__NEED_pthread_condattr_t) && !defined(__DEFINED_pthread_condattr_t)
-typedef struct { unsigned __attr; } pthread_condattr_t;
+typedef struct { union { unsigned __attr;
+#ifdef __aarch64__
+	long __glibc_compat;
+#endif
+};
+} pthread_condattr_t;
 #define __DEFINED_pthread_condattr_t
 #endif

 #if defined(__NEED_pthread_barrierattr_t) && !defined(__DEFINED_pthread_barrierattr_t)
-typedef struct { unsigned __attr; } pthread_barrierattr_t;
+typedef struct { union { unsigned __attr;
+#ifdef __aarch64__
+	long __glibc_compat;
+#endif
+};
+} pthread_barrierattr_t;
 #define __DEFINED_pthread_barrierattr_t
 #endif

@@ -383,12 +398,20 @@


 #if defined(__NEED_pthread_attr_t) && !defined(__DEFINED_pthread_attr_t)
-typedef struct { union { int __i[sizeof(long)==8?14:9]; volatile int __vi[sizeof(long)==8?14:9]; unsigned long __s[sizeof(long)==8?7:9]; } __u; } pthread_attr_t;
+typedef struct { union { int __i[sizeof(long)==8?14:9]; volatile int __vi[sizeof(long)==8?14:9]; unsigned long __s[sizeof(long)==8?7:9];
+#ifdef __aarch64__
+	char __glibc_compat[64];
+#endif
+} __u; } pthread_attr_t;
 #define __DEFINED_pthread_attr_t
 #endif

 #if defined(__NEED_pthread_mutex_t) && !defined(__DEFINED_pthread_mutex_t)
-typedef struct { union { int __i[sizeof(long)==8?10:6]; volatile int __vi[sizeof(long)==8?10:6]; volatile void *volatile __p[sizeof(long)==8?5:6]; } __u; } pthread_mutex_t;
+typedef struct { union { int __i[sizeof(long)==8?10:6]; volatile int __vi[sizeof(long)==8?10:6]; volatile void *volatile __p[sizeof(long)==8?5:6];
+#ifdef __aarch64__
+	char __glibc_compat[48];
+#endif
+} __u; } pthread_mutex_t;
 #define __DEFINED_pthread_mutex_t
 #endif
